{"metadata":{"language_info":{"name":"R","codemirror_mode":"r","pygments_lexer":"r","mimetype":"text/x-r-source","file_extension":".r","version":"3.6.3"},"kernelspec":{"name":"ir","display_name":"R","language":"R"}},"nbformat_minor":5,"nbformat":4,"cells":[{"cell_type":"code","source":"install.packages('bigrquery')\nlibrary(bigrquery)\nlibrary(repr)","metadata":{"trusted":true},"execution_count":1,"outputs":[{"name":"stderr","text":"Installing package into ‘/srv/rlibs’\n(as ‘lib’ is unspecified)\n\nalso installing the dependencies ‘bit’, ‘bit64’, ‘rapidjsonr’\n\n\nRegistered S3 methods overwritten by 'tibble':\n  method     from  \n  format.tbl pillar\n  print.tbl  pillar\n\n","output_type":"stream"}],"id":"829cd2b8-4cee-4aaf-b493-59df23655a38"},{"cell_type":"markdown","source":"# Clinical and cohort queries\nHow old are the patients for each case in a given project?","metadata":{},"id":"429364db-dca1-4277-888c-9632d33a0f38"},{"cell_type":"code","source":"project <- \"isb-cgc-etl\"\n\ncase_sql <- \"select\n  submitter_id case_barcode,\n  demo__age_at_index age,\n  demo__vital_status vital\nfrom `isb-cgc-bq.TCGA.clinical_gdc_current`\nwhere proj__project_id = 'TCGA-BRCA'\n\"\ncase_query <- bq_project_query(project, query=case_sql)\ncase_df <- as.data.frame(bq_table_download(case_query))\nhead(case_df)","metadata":{"trusted":true},"execution_count":null,"outputs":[{"name":"stderr","text":"Waiting for authentication in browser...\n\nPress Esc/Ctrl + C to abort\n\nPlease point your browser to the following url: \n\nhttps://accounts.google.com/o/oauth2/auth?client_id=465736758727.apps.googleusercontent.com&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fbigquery%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform&redirect_uri=http%3A%2F%2Flocalhost%3A1410%2F&response_type=code&state=Lxh8QBfCtU\n\n","output_type":"stream"}],"id":"5994df62-4a09-4c87-a911-7ad060028d68"},{"cell_type":"code","source":"case_num <- length(unique(case_df[,2]))\nhist(case_df[,2], breaks=case_num/2, main='age at index', xlab='age (years)')","metadata":{"trusted":true},"execution_count":null,"outputs":[],"id":"3895cc0a-e53d-4d17-82a0-19bec51429eb"},{"cell_type":"code","source":"hist(case_df[case_df[,3] == 'Alive',2], breaks=case_num/2, xlim=c(25,90), col='steelblue1', main='', xlab='age (years)')\nhist(case_df[case_df[,3] == 'Dead',2], breaks=case_num/2, add=1, col='orange')\nlegend('topright', bty='n', legend=c('Alive', 'Dead'), fill=c('steelblue1', 'orange'))","metadata":{"trusted":true},"execution_count":null,"outputs":[],"id":"ecf3eb7f-a904-48a5-8f17-bb509dbad9a0"},{"cell_type":"markdown","source":"# Gene expressionn\nWhat is the mean RNAseq expression (fpkm-uq) of a specific gene in normal and tumor?\n\nTable: *isb-cgc-bq.TCGA.RNAseq_hg38_gdc_current*","metadata":{},"id":"0fbadc15-3127-4954-943c-38f9402da6e3"},{"cell_type":"code","source":"expr_sql <- \"select\n  case_barcode,\n  aliquot_barcode,\n  sample_type_name,\n  HTSeq__FPKM_UQ\nfrom `isb-cgc-bq.TCGA.RNAseq_hg38_gdc_current`\nwhere project_short_name = 'TCGA-BRCA'\n  and gene_name = 'CCNB1'\"\n\nexpr_query <- bq_project_query(project, query=expr_sql)\nexpr_df <- bq_table_download(expr_query)\nhead(expr_df)","metadata":{"trusted":true},"execution_count":null,"outputs":[],"id":"f4867319-6e3d-4820-bcd4-48bff2b0a82d"},{"cell_type":"code","source":"tumor <- unlist(expr_df[expr_df[,3] == 'Primary Tumor',4])\nnormal <- unlist(expr_df[expr_df[,3] == 'Solid Tissue Normal',4])\nmetastatic <- unlist(expr_df[expr_df[,3] == 'Metastatic',4])\n\nboxplot(list(normal, tumor, metastatic), names=c('normal', 'tumor', 'metastatic'))","metadata":{"trusted":true},"execution_count":null,"outputs":[],"id":"6f4079b9-5136-4059-95cd-68009df2898b"},{"cell_type":"markdown","source":"For larger data sets, instead of downloading these data and computing statistics locally we can save ourselves the compute time and generate summary statistics directly in BigQuery.","metadata":{},"id":"a4fb9974-d5a3-4049-99ae-8377175c5d55"},{"cell_type":"code","source":"summ_sql <- \"select\n  sample_type_name,\n  avg(HTSeq__FPKM_UQ) avg_fpkm,\n  stddev(HTSeq__FPKM_UQ) sdv_fpkm,\n  max(HTSeq__FPKM_UQ) max_fpkm,\n  min(HTSeq__FPKM_UQ) min_fpkm\nfrom `isb-cgc-bq.TCGA.RNAseq_hg38_gdc_current`\nwhere project_short_name = 'TCGA-BRCA'\nand gene_name = 'CCNB1'\ngroup by sample_type_name\"\n\nsumm_query <- bq_project_query(project, query=summ_sql)\nsumm_df <- bq_table_download(summ_query)\nhead(summ_df)","metadata":{"trusted":true},"execution_count":null,"outputs":[],"id":"37b61381-fa8a-4dc4-8c5a-546d47430f3c"},{"cell_type":"markdown","source":"# Genome annotation\nWe can retrieve Gencode annotation data for specific genes from BigQuery tables:\n\nTable: *isb-cgc-bq.GENCODE_versioned.annotation_gtf_hg38_v22*","metadata":{},"id":"c8193a1f-1df6-4ed8-9c3f-608bdd468127"},{"cell_type":"code","source":"annot_sql <- \"select * \nfrom `isb-cgc-bq.GENCODE_versioned.annotation_gtf_hg38_v22` \nwhere gene_name = 'CCNB1' \n  and transcript_name = 'CCNB1-001'\"\nannot_query <- bq_project_query(project, query=annot_sql)\nannot_df <- bq_table_download(annot_query)\nprint(annot_df)","metadata":{"trusted":true},"execution_count":null,"outputs":[],"id":"ab66a3b4-47c5-4e52-b1ab-9f9a470e19d0"},{"cell_type":"markdown","source":"# Mutation data\nA common question in biological research is what known mutations exist in a given gene in other data sets? Given the gene annotation data above we can query for all mutations that occur within a given gene in the TCGA-BRCA cases.\n\nTable: *isb-cgc-bq.TCGA.somatic_mutation_hg38_gdc_current*","metadata":{},"id":"adeca1f5-ceb2-4182-93ca-0a8ebe519914"},{"cell_type":"code","source":"mut_sql <- \"select case_barcode, \n    Chromosome, \n    Start_Position, \n    End_Position, \n    Reference_Allele, \n    Tumor_Seq_Allele1, \n    Tumor_Seq_Allele2\nfrom `isb-cgc-bq.TCGA.somatic_mutation_hg38_gdc_current`\nwhere Chromosome = 'chr5'\n    and Start_Position > 69167135\n    and End_Position < 69177358\"\n\nmut_query <- bq_project_query(project, query=mut_sql)\nmut_df <- bq_table_download(mut_query)\nhead(mut_df)","metadata":{},"execution_count":null,"outputs":[],"id":"df0e4327-e24e-4a47-b4e8-05c107575a95"},{"cell_type":"code","source":"options(repr.plot.width=10, repr.plot.height=4)\nmut_count <- table(mut_df[,3])\nmax_count <- max(mut_count)\nplot(NA, xlim=c(min(mut_df[,3])-1000, max(mut_df[,3])+1000), ylim=c(-0.3,max_count), \n     yaxt='n', bty='n', main='', xlab='Chr5 position', ylab='occurance count')\njnk <- apply(annot_df[annot_df$feature == 'exon',], 1, function(row){\n    rect(row[4], -0.2, row[5], -0.1, col='green')\n})\naxis(2, at=seq(0,3,1))\njnk <- sapply(1:length(mut_count), function(ix){\n    x <- as.numeric(names(mut_count)[ix])\n    segments(x, 0, x, mut_count[ix])\n    })","metadata":{"trusted":true},"execution_count":null,"outputs":[],"id":"a9f9f1a1-c32f-4711-a54d-0ef4670932a0"},{"cell_type":"markdown","source":"What about the regulatory regions of the gene? What are the mutations in the 100 kB upstream and downstream of the gene?","metadata":{},"id":"989f6719-e0f0-40c7-bf07-9d25a01a1417"},{"cell_type":"code","source":"sql <- \"select case_barcode, \n    Chromosome, \n    Start_Position, \n    End_Position, \n    Reference_Allele, \n    Tumor_Seq_Allele1, \n    Tumor_Seq_Allele2\nfrom `isb-cgc-bq.TCGA.somatic_mutation_hg38_gdc_current`\nwhere \n    Chromosome = 'chr9'\n    and (\n        (Start_Position < 69067135 and Start_Position > 69167135)\n        or (End_Position > 69177358 and End_Position < 69277358)\n        )\"\n\nquery <- bq_project_query(project, query=sql)\ndf <- bq_table_download(query)\nhead(df)","metadata":{},"execution_count":null,"outputs":[],"id":"e6476bb8-d28c-4860-a3fd-9301197e3048"},{"cell_type":"markdown","source":"# Table joins\nHaving a given set of gene expression, clinical data, or mutations, how do we join these tables together for more complex analyses such as differential expression analyses or machine learning?\n\nTables: \n*   *isb-cgc-bq.TCGA.clinical_gdc_current*\n*   *isb-cgc-bq.TCGA.somatic_mutation_hg38_gdc_current*","metadata":{},"id":"64d26977-6c22-496e-b541-bf7a531b9cf7"},{"cell_type":"code","source":"sql <- \"select \n    mut.Chromosome, \n    mut.Start_Position, \n    mut.End_Position, \n    mut.Reference_Allele, \n    mut.Tumor_Seq_Allele1, \n    mut.Tumor_Seq_Allele2, \n    clin.demo__vital_status \nfrom `isb-cgc-bq.TCGA.clinical_gdc_current` clin\njoin `isb-cgc-bq.TCGA.somatic_mutation_hg38_gdc_current` mut \n    on clin.submitter_id = mut.case_barcode\nwhere mut.project_short_name = 'TCGA-BRCA'\"\n\nquery <- bq_project_query(project, query=sql)\ndf <- bq_table_download(query)\nhead(df)","metadata":{"trusted":true},"execution_count":null,"outputs":[],"id":"5a274c0a-0a4f-4ffa-aa5d-326aaeb33c32"},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[],"id":"dc164a74-29a2-4972-9429-835b1a1e3260"}]}